Implement RZ for INS with tests.

lindsayad opened the pull request:
- Modified divergence of velocity field in INSMass
- Added counter-balancing pressure term in INSMomentum that compensates for
  the translation of gradient to divergence operator in RZ
- Added two tests to the open_bc test suite. One test integrates the pressure
  by parts, the other does not. The domain inlet has half the radius of the
  domain outlet. This allows testing of whether the RZ formulation conserves
  mass in the presence of non-zero radial velocity. The answer is: it does.

fixes #7651.

Commit:
Add cylindrical coordinate system option to INSMass.

Commit:
Add coordinate type option to INSMomentum and try adding pressure term.

Commit:
Modify INSMass and INSMomentum jacobians (#7651).

Commit:
Implementing RZ for INS with tests (#7651).

- Modified divergence of velocity field in INSMass
- Added counter-balancing pressure term in INSMomentum that compensates for
  the translation of gradient to divergence operator in RZ
- Added two tests to the open_bc test suite. One test integrates the pressure
  by parts, the other does not. The domain inlet has half the radius of the
  domain outlet. This allows testing of whether the RZ formulation conserves
  mass in the presence of non-zero radial velocity. The answer is: it does.
- Add viscous term correction for RZ. This correction comes from a
  non-vanishing term in the theta-theta component of the viscous stress
  tensor.
- Remove unnecessary _v_vel and _w_vel from INSMass.

jwpeterson commented:
The pre-check failed because you added trailing whitespace to the following files: 
```
	modules/navier_stokes/include/kernels/INSMass.h
	modules/navier_stokes/src/kernels/INSMass.C
	modules/navier_stokes/tests/ins/open_bc/RZ_do_nothing_bcs_cone.i
```

lindsayad commented:
Awesome. I thought I had put something in my .emacs for that...

jwpeterson commented on modules/navier_stokes/include/kernels/INSMass.h:
```diff
@@ -27,11 +27,30 @@ class INSMass : public Kernel
+  COORD_TYPE _coord_type;
+
+  // Coupled values
+  const VariableValue & _u_vel;
```
Only `_u_vel` is needed, right?

lindsayad replied:
Correct!

jwpeterson commented:
Have a look at the PDF below.  I derived what I think is the correct weak form for the axisymmetric incompressible Navier-Stokes equations, because Google searching did not lead to anything which was exactly what we needed.  In addition to the two new terms you added, there should be one more term added to the r-momentum equation in Eq. (23).  This term agrees with the thesis of DeParis which I linked earlier, the factor of two comes from the fact that we're using the stress tensor form of the Stokes equations while he was using the Laplacian form...

[cc_stokes.pdf](https://github.com/idaholab/moose/files/446003/cc_stokes.pdf)

jwpeterson commented on modules/navier_stokes/src/kernels/INSMass.C:
```diff
@@ -17,35 +17,77 @@ InputParameters validParams<INSMass>()
   // Note: we (arbitrarily) multilply this term by -1 so that it matches the -p(div v)
   // term in the momentum equation.  Not sure if that is really important?
-  return -(_grad_u_vel[_qp](0) + _grad_v_vel[_qp](1) + _grad_w_vel[_qp](2)) * _test[_i][_qp];
+  switch(_coord_type)
```
Civet is currently failing because you need a space between `switch` and `(_coord_type)`

jwpeterson commented on modules/navier_stokes/src/kernels/INSMass.C:
```diff
@@ -63,13 +105,49 @@ Real INSMass::computeQpJacobian()
   if (jvar == _u_vel_var_number)
-    return -_grad_phi[_j][_qp](0) * _test[_i][_qp];
+  {
+    switch(_coord_type)
```
Also here.

jwpeterson commented on modules/navier_stokes/src/kernels/INSMass.C:
```diff
@@ -63,13 +105,49 @@ Real INSMass::computeQpJacobian()
   else if (jvar == _v_vel_var_number)
-    return -_grad_phi[_j][_qp](1) * _test[_i][_qp];
+  {
+    switch(_coord_type)
```
and here.

jwpeterson commented on modules/navier_stokes/src/kernels/INSMass.C:
```diff
@@ -63,13 +105,49 @@ Real INSMass::computeQpJacobian()
   else if (jvar == _w_vel_var_number)
-    return -_grad_phi[_j][_qp](2) * _test[_i][_qp];
+  {
+    switch(_coord_type)
```
and here.

lindsayad replied:
Haha thanks. I promise I was (am?) about to open the coding standards guide.

jwpeterson commented on modules/navier_stokes/tests/ins/open_bc/RZ_do_nothing_bcs_cone.i:
```diff
@@ -0,0 +1,140 @@
+[Executioner]
+  type = Steady
+  petsc_options_iname = '-pc_type -pc_factor_shift_type -pc_factor_shift_amount -ksp_type'
+  petsc_options_value = 'lu NONZERO 1.e-10 preonly'
```
I'd rather not check in a test which uses `-pc_type lu`.  This is fine for testing, but it means this test would not be able to run in parallel...

lindsayad replied:
Good point. Unfortunately, most of my physics work so far has been with small problems so I've always used direct Newton. I don't have much experience with the best iterative solution options. I know that with `integrate_p_by_parts = true`, this new test does not pass using the options from velocity_inletBC.i

jwpeterson replied:
> On Aug 30, 2016, at 4:22 PM, Alex Lindsay <notifications@github.com> wrote:
> 
> In modules/navier_stokes/tests/ins/open_bc/RZ_do_nothing_bcs_cone.i:
> 
> > +
> > +[Problem]
> > +[]
> > +
> > +[Preconditioning]
> > +  [./SMP_PJFNK]
> > +    type = SMP
> > +    full = true
> > +    solve_type = Newton
> > +  [../]
> > +[]
> > +
> > +[Executioner]
> > +  type = Steady
> > +  petsc_options_iname = '-pc_type -pc_factor_shift_type -pc_factor_shift_amount -ksp_type'
> > +  petsc_options_value = 'lu NONZERO 1.e-10 preonly'
> Good point. Unfortunately, most of my physics work so far has been with small problems so I've always used direct Newton. I don't have much experience with the best iterative solution options. I know that with integrate_p_by_parts = true, this new test does not pass using the options from velocity_inletBC.i
> 

I can take a closer look.  But I'd like to implement the extra term from the PDF I posted first.  Do you want to attempt that, or should I?

lindsayad replied:
I'll do it

lindsayad commented:
When dopey people have to go back and delete trailing whitespace...and then go back and untabify...and then go back and fix their switch spacing...is it preferable to squash silly, small commits like that into one?

jwpeterson commented:
> On Aug 30, 2016, at 4:27 PM, Alex Lindsay <notifications@github.com> wrote:
> 
> When dopey people have to go back and delete trailing whitespace...and then go back and untabify...and then go back and fix their switch spacing...is it preferable to squash silly, small commits like that into one?
> 

Absolutely.

jwpeterson commented:
The new tests seem to be working OK, the debug modules failure is due to unrelated tests timing out.

lindsayad commented:
Yea, planning on looking over your pdf today.

lindsayad commented:
Ok, I added the theta-theta correction. It changed the test results as one would expect. Conservation of mass still happens (less than .01% error now). And I think the u_r profile looks more correct in that it is >= 0 over the entire domain. I don't have any support for that feeling other than intuitively I don't see any reason for momentum to flow back towards the axis of symmetry since the cone is expanding in the +z direction.

![u_r_with_viscous_term_correction](https://cloud.githubusercontent.com/assets/10248304/18153635/6cdb7148-6fc4-11e6-80ee-019e04bdd77a.png)

jwpeterson commented:
The latest image looks good to me, although it seems like you deleted the previous images you had posted (?) so it's tough to compare to the previous results.  Would you mind checking in the Gmsh input file you used to generate 2d_cone.msh?  This will be helpful in case anyone ever wants to change the mesh (we might need to make it coarser so that it runs fast enough for the test harness, etc.)

Design-wise, I think there are still a few things I'd like to change, but I can take care of those before merging:

* Moose already provides an `enum CoordinateSystemType` in MooseTypes.h, so there's no reason to duplicate this via `MooseEnum coord_types("XYZ RZ RSPHERICAL");`.  Also, there should be no need for `setGeometryParameter()` since `FEProblem::getCoordSystem()` can provide this.   
* Along the same lines, I think I'd like to create new RZ-specific Kernels that inherit from the existing `INSMass` and `INSMomentum` Kernels and implement the additional terms.  This should help us avoid having all the case statements in the `computeQpResidual()` and `computeQpJacobian()` functions.  Selecting the proper Kernels for a given simulation would later be simplified by developing an `Action`.  The new tests will help me ensure that nothing gets broken with this refactoring.
* Kernels provide the `_coord[_qp]` type for RZ problems which I believe should be preferred over `_q_point[_qp](0)`, however they are otherwise identical in all cases as far as I can tell.
* The `-pc_type lu` solver option still needs to be addressed before merging.

At this point, my plan is to pull your branch and address the issues above, then provide a PR back to the same branch which you can then merge... you can still push commits here, but please avoid rebasing it any more, so that the history stays in sync.

lindsayad commented:
That's really odd about the graphics. Maybe that happened because I squashed some commits. Here is the old u_r (without the viscous term correction):

![rz_cone_u_no_parts](https://cloud.githubusercontent.com/assets/10248304/18172355/ad1d1796-702a-11e6-98a8-60edbc7856da.png)

Yea I felt silly adding that enum, but I was loosely basing it off of GapConductance in the heat_conduction module. I saw that they could default to _coord_sys which is initialized from _assembly.coordSystem(), but the Kernel class doesn't have that inheritance. However, I didn't know about FEProblem::getCoordSystem! That's definitely better

All your recommended changes sound great. I will push the .geo file. I'm also going to play around a bit with solver options, but you should feel free to do that as well in parallel because you'll probably come up with a solution quicker.

Commit:
Add 2d_cone.geo (#7651).

jwpeterson committed:
Misc. fixes to original branch as discussed in PR.

* Rename velocity components to vel_x, vel_y.
  Paraview picks these names up automatically and converts them into a
  vector, making it easier to apply the Glyph filter.
* Change problem to transient Five timesteps are enough to demonstrate
  the test works..  This now works with the 'standard' solver options
  in both serial and parallel.
* Define a variable for r, avoid std::pow().
  Note: _coord is not the same thing as "r", _coord also includes the 2*pi.
* Regold.

jwpeterson commented:
> All your recommended changes sound great. I will push the .geo file. I'm also going to play around a bit with solver options, but you should feel free to do that as well in parallel because you'll probably come up with a solution quicker.

I think I know what the problem with the solver is.  We haven't pinned a value of the pressure yet in this problem, so the system is indefinite.  (This is also done in the `velocity_inletBC.i` input file with the `p_corner` BC.)  I think I'll probably pin a value on the channel centerline or something and see if that fixes it.  I also changed the name of the velocity vectors in the input file so that it's easier to plot vector fields in Paraview:

![cone_vecs_and_contours](https://cloud.githubusercontent.com/assets/1775907/18173581/639fdf2a-7027-11e6-80be-7f0ca9b36914.png)

jwpeterson commented:
> We haven't pinned a value of the pressure yet in this problem, so the system is indefinite.

Nevermind, just saw the `p_corner` BC in the cone input file.  I wonder if this is somehow not working the way it's supposed to... it's definitely behaving like an indefinite problem...

jwpeterson commented:
Running with ` -pc_type svd -pc_svd_monitor` I've seeming confirmed that the pressure pin is working, there are no nearly zero singular values.
```
 0 Nonlinear |R| = 1.054356e+01
      SVD: condition number 3.265791382127e+07, 0 of 3781 singular values are (nearly) zero
      SVD: smallest singular values: 2.762357777957e-06 6.453557410945e-06 6.746774868554e-06 9.953052248852e-06 1.038787421241e-05
      SVD: largest singular values : 7.352494816447e+01 7.676781730265e+01 8.154027701441e+01 8.248385386976e+01 9.021284225602e+01
 1 Nonlinear |R| = 9.332275e-03
      SVD: condition number 3.009147532926e+07, 0 of 3781 singular values are (nearly) zero
      SVD: smallest singular values: 2.997949204001e-06 6.453510492388e-06 6.750207822765e-06 9.952848485945e-06 1.038786038160e-05
      SVD: largest singular values : 7.352493319368e+01 7.676779927044e+01 8.154010502900e+01 8.248372375855e+01 9.021271451059e+01
 2 Nonlinear |R| = 7.928406e-07
      SVD: condition number 3.018022895632e+07, 0 of 3781 singular values are (nearly) zero
      SVD: smallest singular values: 2.989132665988e-06 6.453499289098e-06 6.750151824734e-06 9.952842864627e-06 1.038786429043e-05
      SVD: largest singular values : 7.352492883271e+01 7.676779325243e+01 8.154010162401e+01 8.248372861002e+01 9.021270824033e+01
 3 Nonlinear |R| = 7.472707e-13
```


I've also confirmed the converse, namely that if I comment out the `p_corner` section, the problem won't solve even with the `-pc_type lu` options in place, and you get 1 nearly zero singular value as expected:
```
 0 Nonlinear |R| = 3.341656e+00
      SVD: condition number 1.229567521603e+19, 1 of 3781 singular values are (nearly) zero
      SVD: smallest singular values: 7.336957155342e-18 6.452541770363e-06 6.715820455307e-06 8.709209505772e-06 9.953976497085e-06
      SVD: largest singular values : 7.352494817980e+01 7.676781730493e+01 8.154027701442e+01 8.248385386984e+01 9.021284225603e+01
```

jwpeterson commented:
So if we add the time derivative terms and use a Transient executioner, the problem converges just fine using the standard solver options in both serial and parallel.  If that's alright with you, I'll just change this to a transient problem, but it's good to know that the steady problem can be solved with LU...

lindsayad commented:
Sounds good to me

jwpeterson commented:
One other thing, you mentioned that you have verified mass conservation by integrating over the inlet and outlet.  Did you write a Postprocessor to do this?  I think it would be a good additional verification feature.

jwpeterson committed:
Add INSMassRZ kernel.

Take RZ stuff back out of INSMass and use INSMassRZ kernel in test.

jwpeterson committed:
Forgot a _mu on the extra term in residual and Jacobian.

lindsayad commented:
No, I exported data as a csv from paraview. Do you want me to add inlet and outlet postprocessors to the test input file?

jwpeterson committed:
Add INSMomentumRZ kernel.

Strip RZ stuff out of INSMomentum kernel and use INSMomentumRZ kernels
in test input file.

jwpeterson commented:
> Do you want me to add inlet and outlet postprocessors to the test input file?

No, I can do it later.  I'm not sure if that's something we already have or if we would need to write code for it.  I've finished with my changes, so I'm going to clean up the branch and make a PR into your branch.  If you merge that PR, then this one should be automatically updated and start testing again...

jwpeterson commented:
> I'm going to clean up the branch and make a PR into your branch.

https://github.com/lindsayad/moose/pull/2

lindsayad committed:
Merge pull request #2 from jwpeterson/RZ_7651

Updates to the Navier-Stokes RZ kernels

lindsayad commented:
Awesome. Thanks for working with me on this @jwpeterson
